### PR TITLE
DYN1805 Validate the performance tracking system by using validated improvements Part II

### PR DESCRIPTION
**Pull request by reddyashish**
### Purpose

This PR is a copy of the one that Craig made: https://github.com/DynamoDS/Dynamo/pull/9620

Instead of removing the TraceUtils class from from Protocore.Lang namespace, we just obsolete it for now. 

Since this trace data is checked for in the Execute function, this part of the code is executed for all the graphs. A small percentage decrease or increase in the run times for those graphs, that take less absolute run times can be ignored. 

Performance results comparison with the base line data(for model tests):
result1
result2

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@mjkkirschner @aparajit-pratap @QilongTang @saintentropy

### CODE REVIEW

**Comment by QilongTang** [comment]
```
Since this trace data is checked for in the Execute function, this part of the code is executed for all the graphs.
```
Can you explain a little bit more or point to the actual caller to help understand?

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 1]
```diff
@@ -17,6 +17,8 @@
 using ProtoCore.Utils;
 using StackFrame = ProtoCore.DSASM.StackFrame;
 using WarningID = ProtoCore.Runtime.WarningID;
+using DynamoServices;
+using Validity = ProtoCore.Utils.Validity;
```
Comment (QilongTang) [inline src/Engine/ProtoCore/Lang/CallSite.cs]: Can you sort the usings? VS should support it by the command `RemoveAndSort`

File: src/Engine/ProtoCore/Lang/TraceUtils.cs [hunk 2]
```diff
@@ -5,6 +5,7 @@
 
 namespace ProtoCore.Lang
 {
+	[Obsolete("This class is not used anymore, use the same class from Dynamo.Services namespace",false)]
     public static class TraceUtils
```
Comment (QilongTang) [inline src/Engine/ProtoCore/Lang/TraceUtils.cs]: alignment?

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 3]
```diff
@@ -1912,16 +1910,18 @@ private FunctionGroup FirstFunctionGroupInInheritanceChain(RuntimeCore runtimeCo
             if (ret.IsNull)
             {
                 //wipe the trace cache
-                TraceUtils.ClearTLSKey(TRACE_KEY);
+                TraceUtils.ClearAllKnownTLSKeys();
+                newTraceData.Data = null;
```
Comment (aparajit-pratap) [inline src/Engine/ProtoCore/Lang/CallSite.cs]: @saintentropy is checking to see if this is required. I pointed this out to him.

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 4]
```diff
@@ -1912,16 +1910,18 @@ private FunctionGroup FirstFunctionGroupInInheritanceChain(RuntimeCore runtimeCo
...
             }
-
```
Comment (reddyashish) [inline src/Engine/ProtoCore/Lang/CallSite.cs]: The cases where ret is not null, it checks for the trace data in the local storage. This happens on every execution of a node.

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 5]
```diff
@@ -1912,16 +1910,18 @@ private FunctionGroup FirstFunctionGroupInInheritanceChain(RuntimeCore runtimeCo
...
-            //TLS -> TraceCache
-            Dictionary<string, ISerializable> traceRet = TraceUtils.GetObjectFromTLS();
-
-            if (traceRet.ContainsKey(TRACE_KEY))
+            else
             {
-                var val = traceRet[TRACE_KEY];
-                newTraceData.Data = val;
+                //TLS -> TraceCache
+                var traceRet = TraceUtils.GetTraceData(TRACE_KEY);
```
Comment (reddyashish) [inline src/Engine/ProtoCore/Lang/CallSite.cs]: traceRet value would be null unless the graph has any trace data

File: src/Engine/ProtoCore/Lang/TraceUtils.cs (hunk 2 above)
Comment (aparajit-pratap) [reply src/Engine/ProtoCore/Lang/TraceUtils.cs]: Also add a TODO to remove this class in 3.0.

Comment (reddyashish) [reply src/Engine/ProtoCore/Lang/TraceUtils.cs]: yes, will align this.

File: src/NodeServices/TraceSupport.cs [hunk 6]
```diff
@@ -17,25 +18,72 @@ public class RegisterForTraceAttribute : Attribute
     /// </summary>
     public static class TraceUtils
     {
+        // ReSharper disable InconsistentNaming
+        //Luke: This is deliberately inconsistent, it is not supposed to be in widespread use, to work around a defiency
+        //in the TLS implementation.
+        //TODO(Luke): Replace this with an attribute lookup
+        internal const string __TEMP_REVIT_TRACE_ID = "{0459D869-0C72-447F-96D8-08A7FB92214B}-REVIT";
```
Comment (QilongTang) [inline src/NodeServices/TraceSupport.cs]: @saintentropy Is this TODO still valid?

Comment (saintentropy) [reply src/NodeServices/TraceSupport.cs]: Not really.  Seems like that ship sailed many years ago

**Comment by QilongTang** [comment]
@reddyashish Is there anything we are still waiting in this PR?

**Comment by reddyashish** [comment]
No Aaron, this is good to go. Merging it.

**Comment by junmendoza** [comment]
Very impressive performance uplift! @aparajit-pratap I'm no longer up to date with the VM but is this improvement attributed to fixing a fundamental issue or an iterative process?

**Comment by aparajit-pratap** [comment]
@junmendoza glad to know you are still tracking progress of Dynamo :) 
This perf improvement was a result of work by @saintentropy : https://github.com/DynamoDS/Dynamo/pull/9620 to replace the backing implementation of TLS  from a thread data slot to a thread relative static field per Microsoft recommendations: https://docs.microsoft.com/en-us/dotnet/standard/threading/thread-local-storage-thread-relative-static-fields-and-data-slots
Since the TLS is checked for each VM interaction it basically affects performance globally, not only for trace/element binding related calls.

**Comment by junmendoza** [comment]
@aparajit-pratap, @saintentropy Interesting! and thanks for the background. Btw did this arise due to performance profiling? Curious how much effort is still put in making the VM faster at this stage :)

**Comment by aparajit-pratap** [comment]
@saintentropy has been slowly chipping away at the VM trying to make performance improvements and he has made some along the way. We have more ongoing work in this area. One of the things I was looking at is at optimizing execution for modified inputs in graphs: https://github.com/DynamoDS/Dynamo/pull/9485. This was based on one of your design proposals some time ago.
This work is yet to be fully completed and merged.

**Pull request merged**